program-runtime/src/invoke_context.rs

```diff
@@ -345,21 +346,24 @@ impl<'a> InvokeContext<'a> {
                     );
                     InstructionError::MissingAccount
                 })?;
-            if let Some(duplicate_index) =
-                deduplicated_instruction_accounts
-                    .iter()
-                    .position(|instruction_account| {
-                        instruction_account.index_in_transaction == index_in_transaction
-                    })
-            {
-                duplicate_indicies.push(duplicate_index);
-                let instruction_account = deduplicated_instruction_accounts
-                    .get_mut(duplicate_index)
-                    .ok_or(InstructionError::NotEnoughAccountKeys)?;
-                instruction_account
-                    .set_is_signer(instruction_account.is_signer() || account_meta.is_signer);
-                instruction_account
-                    .set_is_writable(instruction_account.is_writable() || account_meta.is_writable);
+
+            let index_in_callee = transaction_callee_map
+                .get_mut(index_in_transaction as usize)
+                .ok_or(InstructionError::MissingAccount)?;
+
+            if let Some(index_in_callee) = index_in_callee {
+                let cloned_account = {
+                    let instruction_account = instruction_accounts
+                        .get_mut(*index_in_callee as usize)
+                        .ok_or(InstructionError::NotEnoughAccountKeys)?;
+                    instruction_account
+                        .set_is_signer(instruction_account.is_signer() || account_meta.is_signer);
+                    instruction_account.set_is_writable(
+                        instruction_account.is_writable() || account_meta.is_writable,
+                    );
+                    instruction_account.clone()
```

Comment:
Is there a reason not to inline this `clone()` in the `push()` below?

Reply:
> Is there a reason not to inline this `clone()` in the `push()` below?

You borrow the vector as mutable (in `push`), while borrowing it as immutable (in `clone`).

Reply:
> Also, wouldn't a later setting of the flag leave the earlier occurrences untouched?

I'm using the first occurrence as the source of truth, so that I can check the account permissions directly in the second loop.